src/main/java/org/elasticsearch/common/util/concurrent/PrioritizedEsThreadPoolExecutor.java

```diff
@@ -161,6 +168,39 @@ public int compareTo(PrioritizedRunnable pr) {
             }
             return insertionOrder < ((TieBreakingPrioritizedRunnable) pr).insertionOrder ? -1 : 1;
         }
+
+        public void scheduleTimeout(ScheduledExecutorService timer, final Runnable timeoutCallback, TimeValue timeValue) {
+            this.timeoutCallback = timeoutCallback;
+            synchronized (this) {
+                timeoutFuture = timer.schedule(new Runnable() {
```

Comment:
I think here we need to check if `hasRun` is set already and if so skip the timeout. There is a chance if things are fast that it has already executed, right?